### ISSUE TITLE
Kubelet should enforce access to images after they've already been pulled to a node

### ISSUE THREAD

**Issue by ncdc** [kubernetes/kubernetes#18787]
# Problem statement

Some images require pull secrets in order to be used by the Kubelet. Pull secrets are per namespace. Imagine the following scenario:
1. If a pod in namespaceA has valid pull secrets for `foo/bar:stable`, when it is scheduled to a given node, the Kubelet uses the pull secrets to pull the image to the node
2. If a pod in namespaceB also refers to `foo/bar:stable` but does not have valid pull secrets for it, if it happens to be scheduled to the same node as the pod in step 1, the Kubelet respects the image pull policy
3. Because the image already exists on the node, if the pull policy is Never or IfNotPresent, the Kubelet runs the pod, even though namespaceB shouldn't be able to use that image.
# Proposed solutions
## Option 1 - Kubelet flag to ignore pull policy and always try to pull

Add a flag to the kubelet, `--always-pull-images` (or something similarly named) that, if true, would force the Kubelet to attempt to pull every image prior to running a pod, regardless of the image's pull policy.

Pros:
- Easy to implement
- Guarantees that a namespace is allowed to use an image

Cons:
- Overrides user-specified intent in image pull policy
- Requires remote registry to be online to check authorization
- Introduces small amount of latency to pod startup time (because of the additional network requests to the registry)
## Option 2 - Keep an in-memory cache of which namespaces are allowed to use which images

After an image has been successfully pulled, record the mapping from namespace to image in an in-memory cache in the Kubelet. Subsequent attempts by pods in the same namespace to use the same image would respect the image pull policy. Any attempts by pods in other namespaces to use this image would ignore the image pull policy and instead force a pull attempt.

Pros:
- Doesn't override user-specified intent in image pull policy
- No added latency to pod startup time

Cons:
- Doesn't handle authorization revocation; if the credentials to pull an image have changed in the remote registry, there is no way to evict an entry from the cache, possibly allowing access to an image that would otherwise fail with option 1. This could be mitigated if the entries in the cache have a TTL.
# Suggested path forward

Looking for feedback from the community on this one. I'd like to implement option 1 in the short term, with the default being false (so image pull policy is respected) to keep the status quo. If we can come to an agreement on option 2, I'd like to move to that eventually.

@kubernetes/goog-node @kubernetes/rh-cluster-infra @kubernetes/sig-node

**Comment by dchen1107** [comment]
@ncdc I agreed we should enforce the image access / usage after the image is pulled, and haven't put much thoughts on how to achieve that yet. 

But I don't think option 1 is the right approach here because all cons you listed there, especially the first one: Overrides user-specified intent in image pull policy. Assuming the user specify the pullPolicy as Never or IfNotPresent, they expect when the container is restarted with the same version / tag of the image, the behavior should be the same. The option 1 actually breaks that, and bring a lot of unpredictability to the overall system.

**Comment by ncdc** [comment]
@dchen1107 if you really want to ensure your running images don't change, you should be referring to them by their content-addressable ID. In OpenShift, we have ImageStream resources which store metadata about tags and images, and our resources for performing builds and deployments use the information in the image streams to translate a pull spec such as `foo/bar:latest` into `foo/bar@sha256:...` so that the system knows the exact image that must be used, and you don't face the possibility of a moving tag.

Another option here would be to issue the same HTTP requests to retrieve the image manifest from the registry that would be issued during a real pull, to ensure that the credentials are still valid. If the request succeeds, the container creation can succeed with whatever image already exists on the node for the pull spec in question.

**Comment by ncdc** [comment]
cc @smarterclayton

**Comment by smarterclayton** [comment]
Our concern is that this is seen as a security risk in regular production use of Kubernetes for certain organizations that have what they consider secret images.   We're looking for the most orthogonal solution that gets them over the hump so they aren't blocked by this.  There definitely needs to be a real solution, but I think pull always is similar in many ways to our other kubelet policies.

**Comment by ncdc** [comment]
@dchen1107 would you at least accept a PR that adds a flag to enable pull always, defaulted to false?

**Comment by vishh** [comment]
Is updating user configs to be `PullAlways` an option here? 
Reaching out the registry directly will help with other use cases as well like image size feasibility checks. So I'm in favor of that approach.

**Comment by ncdc** [comment]
@vishh updating in what way? As in hoping the users will do it themselves? Or via an admission controller?

**Comment by vishh** [comment]
Ideally, users update their configs. It that is not an option, may be via a
custom admission controller?

**Comment by ncdc** [comment]
Yeah, we can't rely on users, sadly. If a flag is unpalatable, would you all take this as an admission controller (off by default)?

**Comment by vishh** [comment]
Personally, admission controller seems to be the least *hacky* solution. So
SGTM. cc @davidopp

**Comment by smarterclayton** [comment]
Is having Kubelet policy considered hacky?  What's the difference
between this and saying whether the kubelet accepts privileged pods?

**Comment by vishh** [comment]
@smarterclayton: By changing the config, we are not overriding the API. That was the rationale.

**Comment by davidopp** [comment]
Admission controller sounds fine.

**Comment by smarterclayton** [comment]
We should put that in API conventions or a design doc then.  I think
it's a good principle to hold, but we haven't applied it a lot.

**Comment by ncdc** [comment]
@vishh @davidopp @smarterclayton as @liggitt pointed out to me, if we use an admission controller for this, it overrides the pull policy for each container/image, and we lose the original value. If the cluster admin later decides to disable this admission controller, the pull policies will still be Always. I suppose we could maybe consider storing the original values as annotations, but I'm guessing the better long-term approach will be to implement option 2 from the proposal above.

**Comment by vishh** [comment]
FYI: I'm working on reorganizing image management inside of kubelet. That
work should pave way for implementing Option 2.

**Comment by yujuhong** [comment]
ref: #9620

**Comment by smarterclayton** [comment]
I think that we can fix #41965 by integrating it into this solution:

1. Introduce a central pull secret management cache (pull image first checks secrets are in place)
2. Once secret access has been verified, client certificates could be written to disk (for Docker, in /etc/...).  The pull would require a valid secret.
3. Clean up the client certificate after no one references it.

**Comment by redbaron** [comment]
In duplicated issue I proposed 3rd option which avoids drawback of 

Kubelet should associate images to  serviceaccount. Then apply policy to marked associated images only.

Probably there are more sophisticated ways to associate, like sign it with SA key, but as an example lets just mark it with re-tagging to show proposed flow.

Lets say pod with default/default SA pulls "alpine:3.7", kubelet is doing following:

   * uses "alpine:3.7-default-default" image name as if POD specified exactly that, it then applies imagePullPolicy to this image name
   * if kubelet decides to pull image, it then pulls real image and then retags it to 'alpine:3.7-default-default' and continues with pod startup using retagged name for the rest of the process

This way multi tenancy can be achieved without causing inconvenience to IfNotPresent PODs when registry is not available.

**Comment by duglin** [comment]
What's the status of this one? I just noticed this security hole in some internal testing and it seems like a big issue when trying to run Kube in a multi-tenant environment.

**Comment by duglin** [comment]
/ping @mikebrow 

This is the issue we were discussing yesterday. We're running Kubernetes in a multi-tenant environment and so we'll have multiple users running pods on the same node. This means that if one of them is able to figure out the image name that some other user used then, even if they don't have access to the private repo from which the image was pulled, they'll be able to use the image from the local cache.  Which becomes a security issue.

**Comment by liggitt** [comment]
> What's the status of this one? I just noticed this security hole in some internal testing and it seems like a big issue when trying to run Kube in a multi-tenant environment.

The [`AlwaysPullImages` admission plugin](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#alwayspullimages) is the only current mitigation

**Comment by duglin** [comment]
@liggitt thanks for the pointer!  Sadly, that won't work for us due to the performance impact of a pull each time, but good to know there's a solution for other use cases.

**Comment by liggitt** [comment]
> Sadly, that won't work for us due to the performance impact of a pull each time, but good to know there's a solution for other use cases.

Note that the "pull" does **not** re-pull already-pulled layers, but simply resolves/verifies the manifest  has not changed in the registry (which incidentally requires authenticating to private registries, which enforces the image access). That means in the normal case (where the image has not changed since the last pull), the request size is O(kb).

It *does* put the registry in the critical path of starting a container, since an unavailable registry will fail the manifest check.

**Comment by duglin** [comment]
ooooo that's interesting and great to know - thanks!  That definitely would help a lot, but you are correct that putting the registry check in the critical path (especially for cases where every millisecond matters for container start-up time) is still an issue for us.

**Comment by mikebrow** [comment]
Alpha level fix ships with v.1.33 :-)
